### PR TITLE
👌 IMPROVE: Alternative proposal for allowing colon fences

### LINKED ISSUE #13: colon fence syntax is broken by escape character
### Describe the problem

In case of a colon fence:

```
:::{NOTE}

README.md should contain the component or package name in it's title.
Using only README as title will lead to lots of "README" items in the outline of the generated documentation.
:::
```

mdformat produces:

```
:::\{NOTE}

README.md should contain the component or package name in it's title.
Using only README as title will lead to lots of "README" items in the outline of the generated documentation.
:::
```

Ideally, mdformat with mdformat_myst would respect this syntax and not escape it.

To me, it is not even clear why mdformat escapes curly brackets in the first place.

### Link to your repository or website

_No response_

### Steps to reproduce

1. "pip install mdformat_myst"
1. Put the above example in a .md file
1. run "mdformat filename.md"

### The version of Python you're using

3.8

### Your operating system

Ubuntu 20.04

### Versions of your packages

```
mdformat==0.7.10
mdformat_frontmatter==0.4.1
mdformat_myst==0.1.4
mdformat_tables==0.4.1
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by brunobeltran**
An alternative proposal to #36 (fixes #13, thanks @nthiery for your pass at this, which inspired me to give it a shot).

Avoids the need for an explicit list of "container_names" (which doesn't seem like an appropriate thing for mdformat_myst to manage) by instead recursively calling mdformat's API directly. I'm not sure if there are any negative consequences to this, as I'm not very knowledgable on the mdformat codebase, but it doesn't seem to lead to a performance regression on $DAY_JOB's monorepo, at least. 

Note that this PR also proposes a very noticeably different set of semantics for when to add padding newlines, namely: I only add padding newlines around _child_ colon fences, so "regular" (e.g. paragraph) content does not get additional padding, maintaining more similarity with the current behavior.

### CODE REVIEW

**Comment by brunobeltran** [comment]
@KyleKing are you still the right person to review this? Sorry for the inbox-blast if not (and please let me know if there's a better comms channel than just GitHub comments for this)!

**Comment by nthiery** [comment]
Great; thank you @brunobeltran  for taking over ;-)

File: tests/data/fixtures.md [hunk 1]
```diff
@@ -504,3 +504,130 @@ MyST directive, no opts or content
...
+- Item with directive
+  :::{tip} Nested tip in list item
+  Tip content inside a list item.
+  :::
+.
+- Item with directive
+  :::{tip} Nested tip in list item
+  Tip content inside a list item.
+  :::
+.
```
Comment (KyleKing) [inline tests/data/fixtures.md:633]: If not already added, can you make sure that other edge cases are tested? For example: with an empty directive, with arguments, containing a code block and deeper nesting, and maybe one to test that regular content with ::: is handled

Additionally, are any tests from #36 not covered here?

File: mdformat_myst/plugin.py [hunk 2]
```diff
@@ -144,15 +140,17 @@ def _escape_text(text: str, node: RenderTreeNode, context: RenderContext) -> str
     return text
 
 
+CHANGES_AST = True
```
Comment (KyleKing) [inline mdformat_myst/plugin.py:143]: Is this strictly necessary? Turning this on can introduce a lot of risk -- I'm assuming this is required to deal with the recursion?

File: mdformat_myst/plugin.py [hunk 3]
```diff
@@ -17,23 +18,16 @@
...
+    plugins_to_enable = [
+        "tables",
+        "front_matters",
+        "footnote",
+    ]
+    for plugin_name in plugins_to_enable:
+        plugin = mdformat.plugins.PARSER_EXTENSIONS[plugin_name]
+        if plugin not in mdit.options["parser_extension"]:
+            mdit.options["parser_extension"].append(plugin)
+            plugin.update_mdit(mdit)
```
Comment (KyleKing) [inline mdformat_myst/plugin.py:30]: Nice! 🌟

File: mdformat_myst/_directives.py [hunk 4]
```diff
@@ -59,49 +73,68 @@ def fence(node: "RenderTreeNode", context: "RenderContext") -> str:
...
+        # Resolve modules back to their string names
+        # mdformat.text() requires names (str), not objects
+        extension_names = [
+            name
+            for name, plugin in mdformat.plugins.PARSER_EXTENSIONS.items()
+            if plugin in active_plugins
+        ]
+        formatted += mdformat.text(
+            content, options=context.options, extensions=extension_names
+        )
```
Comment (KyleKing) [inline mdformat_myst/_directives.py:128]: I'm a little worried about this the performance here and would prefer to have native support like regex parsing (proposed in #36), but directive content should be small and is something we could revisit

**Review by KyleKing — APPROVED** [review]
Some minor comments and overall looks like this should work with some caveats on performance. Thanks!

File: mdformat_myst/_directives.py (hunk 4 above)
Comment (KyleKing) [reply mdformat_myst/_directives.py:128]: The bigger issue is that I don't think any other plugin uses mdformat recursively like this, so it could lead to unexpected behavior

**Comment by njzjz** [comment]
The mdformat in this PR formats the following markdown

````md
```{eval-rst}
.. dargs::
   :module: deepmd.utils.argcheck
   :func: gen_args_multi_task
```
````

to

````md
```{eval-rst}
.. dargs::
:module: deepmd.utils.argcheck
:func: gen_args_multi_task
```
````

The spaces are removed incorrectly.
